src/file-events.c

```diff
@@ -0,0 +1,232 @@
+  .key_size = sizeof(u64),
+  .value_size = sizeof(incomplete_mkdir_t),
+  .max_entries = 256,
+  .pinning = 0,
+  .namespace = "",
+};
+
+static __always_inline void enter_mkdir(struct pt_regs *ctx)
+{
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    incomplete_mkdir_t event = {0};
+    event.pid_tgid = pid_tgid;
+    event.start_ktime_ns = bpf_ktime_get_ns();
+
+    int ret = bpf_map_update_elem(&incomplete_mkdirs, &pid_tgid, &event, BPF_ANY);
+    if (ret < 0)
+    {
+        file_message_t fm = {0};
+        fm.type = FM_WARNING;
+        fm.u.warning.pid_tgid = pid_tgid;
+        fm.u.warning.message_type.file = FM_CREATE;
+        fm.u.warning.code = W_UPDATE_MAP_ERROR;
+        fm.u.warning.info.err = ret;
+
+        push_file_message(ctx, &fm);
+
+        return;
+    }
+
+    return;
+}
+
+static __always_inline void store_dentry(struct pt_regs *ctx, void *dentry)
+{
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+
+    load_event(incomplete_mkdirs, pid_tgid, incomplete_mkdir_t);
+    if (event.target_dentry == NULL) {
+        event.target_dentry = dentry;
+        bpf_map_update_elem(&incomplete_mkdirs, &pid_tgid, &event, BPF_ANY);
+    }
+    return;
+
+    EventMismatch:;
+    file_message_t fm = {0};
+    fm.type = FM_WARNING;
+    fm.u.warning.pid_tgid = pid_tgid;
+    fm.u.warning.message_type.file = FM_CREATE;
+    fm.u.warning.code = W_PID_TGID_MISMATCH;
+    fm.u.warning.info.stored_pid_tgid = event.pid_tgid;
+
+    push_file_message(ctx, &fm);
+    return;
+
+    NoEvent:
+    return;
+}
+
+static __always_inline void exit_mkdir(struct pt_regs *ctx, tail_call_slot_t tail_call)
+{
+    u32 key = 0;
+    buf_t *buffer = (buf_t *)bpf_map_lookup_elem(&buffers, &key);
+    if (buffer == NULL) return;
+
+    cached_path_t *cached_path = (cached_path_t *)bpf_map_lookup_elem(&percpu_path, &key);
+    if (cached_path == NULL) return;
+
+    file_message_t *fm = (file_message_t *)buffer;
+    cursor_t cursor = { .buffer = buffer, .offset = &fm->u.action.buffer_len };
+    int ret = 0;
+    error_info_t info = {0};
+
+    if (cached_path->next_dentry != NULL) goto ResolveName;
+
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    load_event(incomplete_mkdirs, pid_tgid, incomplete_mkdir_t);
+    ret = extract_file_info_owner(event.target_dentry, &fm->u.action.target, &fm->u.action.target_owner);
+    if (ret < 0) goto EmitWarning;
+
+    fm->type = FM_CREATE;
+    fm->u.action.pid = event.pid_tgid >> 32;
+    fm->u.action.mono_ns = event.start_ktime_ns;
+    fm->u.action.buffer_len = sizeof(file_message_t);
+    fm->u.action.u.create.source_link = LINK_NONE;
+
+    cached_path->next_dentry = event.target_dentry;
+
+    ResolveName:
+    ret = write_path(ctx, cached_path, &cursor, tail_call);
+    cached_path->next_dentry = NULL;
+    if (ret < 0) goto EmitWarning;
+
+    write_null_char(cursor.buffer, cursor.offset);
+    push_flexible_file_message(ctx, fm, *cursor.offset);
+    return;
+
+    EventMismatch:
```

Comment:
question: Does this need to be `EventMismatch:;` (end with semicolon)?

Reply:
It's not normal to put a semicolon after a label; I think it was done in some places to allow a variable declaration immediately after the label, which isn't normally allowed.